docs/rules/no-mutation-props.md

```diff
@@ -0,0 +1,63 @@
+# Prevent mutation of this.props (no-mutation-props)
+
+NEVER mutate `this.props`, as all React components must act like pure functions with respect to their props. 
+Treat `this.props` as if it were immutable. More info available at [https://facebook.github.io/react/docs/components-and-props.html#props-are-read-only](https://facebook.github.io/react/docs/components-and-props.html#props-are-read-only)
+
+## Rule Details
+
+This rule is aimed to forbid the use of mutating `this.props`.
+
+The following patterns are considered warnings:
+
+```jsx
+var Hello = React.createClass({
+  render: function() {
+    this.props.name = this.props.name.toUpperCase();
+    return <div>Hello {this.props.name}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    const {list} = this.props;
+    list.push(2);
```

Comment:
i don't see how this can safely be considered a mutation; `list` could be anything with a "push" method, there's no way to know it's an array.

Reply:
That's a great point. I'd hate to loose the functionality though. Any thoughts on how we could determine it's an array?

Reply:
Meanwhile, [on Gitter](https://gitter.im/eslint/eslint?at=59b1e7a8bac826f05490b79b)… 

@not-an-aardvark suggests that leave this rule as-is since these method names tend to imply mutation anyway (also, it would be pretty weird to have a `prop` that isn't a plain object). Further, people can disable custom use-cases with an inline `eslint-disable`.

If we want to be very safe, we could put this behind an option flag?

Reply:
I think putting it behind an option is a great compromise - what about two booleans, and two lists of 1) instance methods, and 2) static methods? That way the instance method list could default to `ARRAY_MUTATIONS`, and the static method list could default to `['Object.assign', 'Object.defineProperties', 'Object.defineProperty']` - and both booleans would default to `false`.

That way most people wouldn't touch the lists at all, they'd just enable the booleans - but those who wanted to update the lists could do so.

Reply:
I'm with you on the array mutations, but I'm not sure I understand the danger of the static methods. Can you run that by me one more time? When is `Object.assign(this.props, {…})` ever a safe operation?